<style type="text/css">
    tr.d0 td {background-color: oldlace; color: black;}
    tr.d1 td {background-color: aliceblue; color: black;}
    </style><?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Galaxy rgPicardASMetrics.py tool output - see http://g2.trac.bx.psu.edu/" />
<title></title>
<link rel="stylesheet" href="/static/style/base.css" type="text/css" />
</head>
<body>
<div class="document">
<h3><a href="http://rgenetics.org">Rgenetics</a> tool rgPicardASMetrics.py run at 02/11/2010 15:49:25</h3><b>Your job produced the following output files.</b><hr/>
<table>
<tr><td><a href="rgPicardASMetrics.out">rgPicardASMetrics.out</a></td></tr>
<tr><td><a href="Alignment_Summary_Metrics.log">Alignment_Summary_Metrics.log</a></td></tr>
</table><p/>
<b>Picard output (transposed for readability)</b><hr/>
<table cellpadding="3" >
<tr class="d0"><td>## net.sf.picard.metrics.StringHeader</td></tr><tr class="d1"><td># net.sf.picard.analysis.CollectAlignmentSummaryMetrics INPUT=/home/raphael/code/channing/galaxy/test-data/picard_input_tiny.sam OUTPUT=/tmp/AsMetrics/rgPicardASMetrics.out REFERENCE_SEQUENCE=/home/raphael/code/channing/galaxy/test-data/picard_input_hg18.trimmed.fasta ASSUME_SORTED=true MAX_INSERT_SIZE=100000 ADAPTER_SEQUENCE=[AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT, AGATCGGAAGAGCTCGTATGCCGTCTTCTGCTTG, AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT, AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG, AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT, AGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNNNATCTCGTATGCCGTCTTCTGCTTG, IS_BISULFITE_SEQUENCED=false] TMP_DIR=/tmp VALIDATION_STRINGENCY=LENIENT    IS_BISULFITE_SEQUENCED=false VERBOSITY=INFO QUIET=false COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false</td></tr><tr class="d0"><td>## net.sf.picard.metrics.StringHeader</td></tr><tr class="d1"><td># Started on: Tue Nov 02 15:49:25 EDT 2010</td></tr><tr class="d0"><td>## METRICS CLASS	net.sf.picard.analysis.AlignmentSummaryMetrics</td></tr><tr class="d1"><td>CATEGORY</td><td>FIRST_OF_PAIR</td></tr>
<tr class="d0"><td>TOTAL_READS</td><td>4</td></tr>
<tr class="d1"><td>PF_READS</td><td>4</td></tr>
<tr class="d0"><td>PCT_PF_READS</td><td>1</td></tr>
<tr class="d1"><td>PF_NOISE_READS</td><td>0</td></tr>
<tr class="d0"><td>PF_READS_ALIGNED</td><td>4</td></tr>
<tr class="d1"><td>PCT_PF_READS_ALIGNED</td><td>1</td></tr>
<tr class="d0"><td>PF_HQ_ALIGNED_READS</td><td>4</td></tr>
<tr class="d1"><td>PF_HQ_ALIGNED_BASES</td><td>404</td></tr>
<tr class="d0"><td>PF_HQ_ALIGNED_Q20_BASES</td><td>28</td></tr>
<tr class="d1"><td>PF_HQ_MEDIAN_MISMATCHES</td><td>82</td></tr>
<tr class="d0"><td>PF_HQ_ERROR_RATE</td><td>0.811881</td></tr>
<tr class="d1"><td>MEAN_READ_LENGTH</td><td>101</td></tr>
<tr class="d0"><td>READS_ALIGNED_IN_PAIRS</td><td>3</td></tr>
<tr class="d1"><td>PCT_READS_ALIGNED_IN_PAIRS</td><td>0.75</td></tr>
<tr class="d0"><td>BAD_CYCLES</td><td>101</td></tr>
<tr class="d1"><td>STRAND_BALANCE</td><td>0.25</td></tr>
<tr class="d0"><td>PCT_CHIMERAS</td><td>0</td></tr>
<tr class="d1"><td>PCT_ADAPTER
</td><td>0
</td></tr>
</table>
<b>Picard log</b><hr/>
<pre>[Tue Nov 02 15:49:25 EDT 2010] net.sf.picard.analysis.CollectAlignmentSummaryMetrics INPUT=/home/raphael/code/channing/galaxy/test-data/picard_input_tiny.sam OUTPUT=/tmp/AsMetrics/rgPicardASMetrics.out REFERENCE_SEQUENCE=/home/raphael/code/channing/galaxy/test-data/picard_input_hg18.trimmed.fasta ASSUME_SORTED=true MAX_INSERT_SIZE=100000 ADAPTER_SEQUENCE=[AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT, AGATCGGAAGAGCTCGTATGCCGTCTTCTGCTTG, AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT, AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG, AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT, AGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNNNATCTCGTATGCCGTCTTCTGCTTG, IS_BISULFITE_SEQUENCED=false] TMP_DIR=/tmp VALIDATION_STRINGENCY=LENIENT    IS_BISULFITE_SEQUENCED=false VERBOSITY=INFO QUIET=false COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false
WARNING	2010-11-02 15:49:25	CollectAlignmentSummaryMetrics	May not be able collect summary statistics in file /home/raphael/code/channing/galaxy/test-data/picard_input_tiny.sam because it is not sorted in coordinate order.  If any of the reads are aligned this will blow up.
[Tue Nov 02 15:49:25 EDT 2010] net.sf.picard.analysis.CollectAlignmentSummaryMetrics done.
Runtime.totalMemory()=46923776
</pre><hr/>The freely available <a href="http://picard.sourceforge.net/command-line-overview.shtml">Picard software</a> 
generated all outputs reported here, using this command line:<br/>
<pre>java -Xmx2g -jar /home/raphael/code/channing/galaxy/tool-data/shared/jars/CollectAlignmentSummaryMetrics.jar REFERENCE_SEQUENCE=/home/raphael/code/channing/galaxy/test-data/picard_input_hg18.trimmed.fasta ASSUME_SORTED=true ADAPTER_SEQUENCE= IS_BISULFITE_SEQUENCED=false MAX_INSERT_SIZE=100000 INPUT=/home/raphael/code/channing/galaxy/test-data/picard_input_tiny.sam OUTPUT=/tmp/AsMetrics/rgPicardASMetrics.out VALIDATION_STRINGENCY=LENIENT TMP_DIR=/tmp</pre>
</div></body></html>

