<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Galaxy rgQC.py tool output - see http://g2.trac.bx.psu.edu/" />
<title></title>
<link rel="stylesheet" href="/static/style/base.css" type="text/css" />
</head>
<body>
<div class="document">
<h4>
<div>Output from Rgenetics QC report tool run at 19/05/2010 15:15:46<br>
</h4>
</div><div><h4>(Click any preview image to download a full sized PDF version)</h4><br><ol>
<table cellpadding="5" border="0">
<tr><td colspan="3"><a href="MarkerDetails_rgQCtest1.xls" type="application/vnd.ms-excel">Click here to download the Marker QC Detail report file</a> (1.4 KB) tab delimited</td></tr>
<tr><td colspan="3"><a href="SubjectDetails_rgQCtest1.xls" type="application/vnd.ms-excel">Click here to download the Subject QC Detail report file</a> (1.4 KB) tab delimited</td></tr>
<tr><td><a href="tinywga_All_Paged.pdf"><img src="tinywga_All_Paged-0.jpg" alt="All tinywga QC Plots joined into a single pdf" hspace="10" align="middle">
</a></td><td>All tinywga QC Plots joined into a single pdf</td><td>&nbsp;</td></tr>
<tr><td><a href="tinywga_All_3x3.pdf"><img src="tinywga_All_3x3-0.jpg" alt="All tinywga QC Plots 3 by 3 to a page" hspace="10" align="middle">
</a></td><td>All tinywga QC Plots 3 by 3 to a page</td><td>&nbsp;</td></tr>
<tr><td><a href="tinywga_logphweunaff.pdf"><img src="tinywga_logphweunaff.jpg" alt="Marker HWE" hspace="10" align="middle">
</a></td><td>Marker HWE</td><td><a href="Ranked_Marker_HWE.xls">Worst data</a></td></tr>
<tr><td><a href="tinywga_logphweunaff_cum.pdf"><img src="tinywga_logphweunaff_cum.jpg" alt="Ranked Marker HWE" hspace="10" align="middle">
</a></td><td>Ranked Marker HWE</td><td>&nbsp;</td></tr>
<tr><td><a href="QQtinywga_logphweunaff_cum.pdf"><img src="QQtinywga_logphweunaff_cum.jpg" alt="LogQQ plot Marker HWE" hspace="10" align="middle">
</a></td><td>LogQQ plot Marker HWE</td><td>&nbsp;</td></tr>
<tr><td><a href="tinywga_missfrac.pdf"><img src="tinywga_missfrac.jpg" alt="Marker Missing Genotype" hspace="10" align="middle">
</a></td><td>Marker Missing Genotype</td><td><a href="Ranked_Marker_Missing_Genotype.xls">Worst data</a></td></tr>
<tr><td><a href="tinywga_missfrac_cum.pdf"><img src="tinywga_missfrac_cum.jpg" alt="Ranked Marker Missing Genotype" hspace="10" align="middle">
</a></td><td>Ranked Marker Missing Genotype</td><td>&nbsp;</td></tr>
<tr><td><a href="tinywga_maf.pdf"><img src="tinywga_maf.jpg" alt="Marker MAF" hspace="10" align="middle">
</a></td><td>Marker MAF</td><td><a href="Ranked_Marker_MAF.xls">Worst data</a></td></tr>
<tr><td><a href="tinywga_maf_cum.pdf"><img src="tinywga_maf_cum.jpg" alt="Ranked Marker MAF" hspace="10" align="middle">
</a></td><td>Ranked Marker MAF</td><td>&nbsp;</td></tr>
<tr><td><a href="tinywga_fracmiss.pdf"><img src="tinywga_fracmiss.jpg" alt="Subject Missing Genotype" hspace="10" align="middle">
</a></td><td>Subject Missing Genotype</td><td><a href="Ranked_Subject_Missing_Genotype.xls">Worst data</a></td></tr>
<tr><td><a href="tinywga_fracmiss_cum.pdf"><img src="tinywga_fracmiss_cum.jpg" alt="Ranked Subject Missing Genotype" hspace="10" align="middle">
</a></td><td>Ranked Subject Missing Genotype</td><td>&nbsp;</td></tr>
<tr><td><a href="tinywga_s_het.pdf"><img src="tinywga_s_het.jpg" alt="Subject F Statistic" hspace="10" align="middle">
</a></td><td>Subject F Statistic</td><td>&nbsp;</td></tr>
<tr><td><a href="tinywga_s_het_cum.pdf"><img src="tinywga_s_het_cum.jpg" alt="Ranked Subject F Statistic" hspace="10" align="middle">
</a></td><td>Ranked Subject F Statistic</td><td>&nbsp;</td></tr>
<tr><td><a href="FQNormtinywga_s_het_cum.pdf"><img src="FQNormtinywga_s_het_cum.jpg" alt="F Statistic Subject F Statistic" hspace="10" align="middle">
</a></td><td>F Statistic Subject F Statistic</td><td>&nbsp;</td></tr>
</table><hr><h3>All output files from the QC run are available below</h3>
<table cellpadding="5" border="0">

<tr><td><a href="FQNormtinywga_s_het_cum.jpg">FQNormtinywga_s_het_cum.jpg</a> (30.2 KB)</td></tr>
<tr><td><a href="FQNormtinywga_s_het_cum.pdf">FQNormtinywga_s_het_cum.pdf</a> (7.1 KB)</td></tr>
<tr><td><a href="MarkerDetails_rgQCtest1.xls">MarkerDetails_rgQCtest1.xls</a> (1.4 KB)</td></tr>
<tr><td><a href="QQtinywga_logphweunaff_cum.jpg">QQtinywga_logphweunaff_cum.jpg</a> (33.0 KB)</td></tr>
<tr><td><a href="QQtinywga_logphweunaff_cum.pdf">QQtinywga_logphweunaff_cum.pdf</a> (6.6 KB)</td></tr>
<tr><td><a href="Ranked_Marker_HWE.xls">Ranked_Marker_HWE.xls</a> (763 B)</td></tr>
<tr><td><a href="Ranked_Marker_MAF.xls">Ranked_Marker_MAF.xls</a> (564 B)</td></tr>
<tr><td><a href="Ranked_Marker_Missing_Genotype.xls">Ranked_Marker_Missing_Genotype.xls</a> (659 B)</td></tr>
<tr><td><a href="Ranked_Subject_Missing_Genotype.xls">Ranked_Subject_Missing_Genotype.xls</a> (804 B)</td></tr>
<tr><td><a href="SubjectDetails_rgQCtest1.xls">SubjectDetails_rgQCtest1.xls</a> (1.4 KB)</td></tr>
<tr><td><a href="ldp_tinywga.bed">ldp_tinywga.bed</a> (143 B)</td></tr>
<tr><td><a href="ldp_tinywga.bim">ldp_tinywga.bim</a> (392 B)</td></tr>
<tr><td><a href="ldp_tinywga.fam">ldp_tinywga.fam</a> (603 B)</td></tr>
<tr><td><a href="ldp_tinywga.log">ldp_tinywga.log</a> (2.0 KB)</td></tr>
<tr><td><a href="tinywga.fmendel">tinywga.fmendel</a> (378 B)</td></tr>
<tr><td><a href="tinywga.frq">tinywga.frq</a> (1.2 KB)</td></tr>
<tr><td><a href="tinywga.het">tinywga.het</a> (2.5 KB)</td></tr>
<tr><td><a href="tinywga.hwe">tinywga.hwe</a> (6.5 KB)</td></tr>
<tr><td><a href="tinywga.imendel">tinywga.imendel</a> (612 B)</td></tr>
<tr><td><a href="tinywga.imiss">tinywga.imiss</a> (1.9 KB)</td></tr>
<tr><td><a href="tinywga.lmendel">tinywga.lmendel</a> (571 B)</td></tr>
<tr><td><a href="tinywga.lmiss">tinywga.lmiss</a> (1.1 KB)</td></tr>
<tr><td><a href="tinywga.log">tinywga.log</a> (1.9 KB)</td></tr>
<tr><td><a href="tinywga.mendel">tinywga.mendel</a> (56 B)</td></tr>
<tr><td><a href="tinywga.prune.in">tinywga.prune.in</a> (140 B)</td></tr>
<tr><td><a href="tinywga.prune.out">tinywga.prune.out</a> (110 B)</td></tr>
<tr><td><a href="tinywga.sexcheck">tinywga.sexcheck</a> (2.5 KB)</td></tr>
<tr><td><a href="tinywga_All_3x3-0.jpg">tinywga_All_3x3-0.jpg</a> (45.2 KB)</td></tr>
<tr><td><a href="tinywga_All_3x3-1.jpg">tinywga_All_3x3-1.jpg</a> (19.3 KB)</td></tr>
<tr><td><a href="tinywga_All_3x3.pdf">tinywga_All_3x3.pdf</a> (82.4 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-0.jpg">tinywga_All_Paged-0.jpg</a> (13.6 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-1.jpg">tinywga_All_Paged-1.jpg</a> (26.3 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-10.jpg">tinywga_All_Paged-10.jpg</a> (31.4 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-11.jpg">tinywga_All_Paged-11.jpg</a> (29.8 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-2.jpg">tinywga_All_Paged-2.jpg</a> (32.3 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-3.jpg">tinywga_All_Paged-3.jpg</a> (11.1 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-4.jpg">tinywga_All_Paged-4.jpg</a> (26.7 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-5.jpg">tinywga_All_Paged-5.jpg</a> (18.4 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-6.jpg">tinywga_All_Paged-6.jpg</a> (28.1 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-7.jpg">tinywga_All_Paged-7.jpg</a> (11.9 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-8.jpg">tinywga_All_Paged-8.jpg</a> (29.6 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged-9.jpg">tinywga_All_Paged-9.jpg</a> (17.9 KB)</td></tr>
<tr><td><a href="tinywga_All_Paged.pdf">tinywga_All_Paged.pdf</a> (81.9 KB)</td></tr>
<tr><td><a href="tinywga_fracmiss.jpg">tinywga_fracmiss.jpg</a> (12.0 KB)</td></tr>
<tr><td><a href="tinywga_fracmiss.pdf">tinywga_fracmiss.pdf</a> (6.4 KB)</td></tr>
<tr><td><a href="tinywga_fracmiss_cum.jpg">tinywga_fracmiss_cum.jpg</a> (30.0 KB)</td></tr>
<tr><td><a href="tinywga_fracmiss_cum.pdf">tinywga_fracmiss_cum.pdf</a> (7.0 KB)</td></tr>
<tr><td><a href="tinywga_logphweunaff.jpg">tinywga_logphweunaff.jpg</a> (13.6 KB)</td></tr>
<tr><td><a href="tinywga_logphweunaff.pdf">tinywga_logphweunaff.pdf</a> (5.7 KB)</td></tr>
<tr><td><a href="tinywga_logphweunaff_cum.jpg">tinywga_logphweunaff_cum.jpg</a> (26.8 KB)</td></tr>
<tr><td><a href="tinywga_logphweunaff_cum.pdf">tinywga_logphweunaff_cum.pdf</a> (5.8 KB)</td></tr>
<tr><td><a href="tinywga_maf.jpg">tinywga_maf.jpg</a> (18.5 KB)</td></tr>
<tr><td><a href="tinywga_maf.pdf">tinywga_maf.pdf</a> (6.4 KB)</td></tr>
<tr><td><a href="tinywga_maf_cum.jpg">tinywga_maf_cum.jpg</a> (28.5 KB)</td></tr>
<tr><td><a href="tinywga_maf_cum.pdf">tinywga_maf_cum.pdf</a> (6.0 KB)</td></tr>
<tr><td><a href="tinywga_missfrac.jpg">tinywga_missfrac.jpg</a> (11.1 KB)</td></tr>
<tr><td><a href="tinywga_missfrac.pdf">tinywga_missfrac.pdf</a> (5.9 KB)</td></tr>
<tr><td><a href="tinywga_missfrac_cum.jpg">tinywga_missfrac_cum.jpg</a> (27.3 KB)</td></tr>
<tr><td><a href="tinywga_missfrac_cum.pdf">tinywga_missfrac_cum.pdf</a> (5.9 KB)</td></tr>
<tr><td><a href="tinywga_s_het.jpg">tinywga_s_het.jpg</a> (18.0 KB)</td></tr>
<tr><td><a href="tinywga_s_het.pdf">tinywga_s_het.pdf</a> (6.8 KB)</td></tr>
<tr><td><a href="tinywga_s_het_cum.jpg">tinywga_s_het_cum.jpg</a> (31.9 KB)</td></tr>
<tr><td><a href="tinywga_s_het_cum.pdf">tinywga_s_het_cum.pdf</a> (7.3 KB)</td></tr>
</table><h4>QC run log contents</h4><pre>## subject reports starting at 19/05/2010 15:15:38
## imissfile /opt/galaxy/test-data/rgtestouts/rgQC/tinywga.imiss contained 40 ids
### writing /opt/galaxy/test-data/rgtestouts/rgQC/SubjectDetails_rgQCtest1.xls report with ['famId', 'iId', 'FracMiss', 'Mendel_errors', 'Ped_sex', 'SNP_sex', 'Status', 'XHomEst', 'F_Stat']## marker reports starting at 19/05/2010 15:15:38
hwe header testpos=2,ppos=8,snppos=1
## starting plotpage, newfpath=/opt/galaxy/test-data/rgtestouts/rgQC,m=[['snp', 'chromosome', 'offset', 'maf', 'a1', 'a2', 'missfrac', 'p_hwe_all', 'logp_hwe_all', 'p_hwe_unaff', 'logp_hwe_unaff', 'N_Mendel'], ['rs2283802', '22', '21784722', '0.2593', '4', '2', '0', '0.638', '0.195179', '0.638', '0.195179', '0']],s=[['famId', 'iId', 'FracMiss', 'Mendel_errors', 'Ped_sex', 'SNP_sex', 'Status', 'XHomEst', 'F_Stat'], ['101', '1', '0.04', '0', '2', '0', 'PROBLEM', 'nan', '-0.03355']]/n## Rgenetics: http://rgenetics.org Galaxy Tools rgQC.py Plink runner

@----------------------------------------------------------@
|        PLINK!       |     v1.06      |   24/Apr/2009     |
|----------------------------------------------------------|
|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |
|----------------------------------------------------------|
|  For documentation, citation & bug-report instructions:  |
|        http://pngu.mgh.harvard.edu/purcell/plink/        |
@----------------------------------------------------------@

Skipping web check... [ --noweb ] 
Writing this text to log file [ tinywga.log ]
Analysis started: Wed May 19 15:15:38 2010

Options in effect:
	--noweb
	--out tinywga
	--bfile /opt/galaxy/test-data/tinywga
	--mind 1.0
	--geno 1.0
	--maf 0.0
	--freq

Reading map (extended format) from [ /opt/galaxy/test-data/tinywga.bim ] 
25 markers to be included from [ /opt/galaxy/test-data/tinywga.bim ]
Reading pedigree information from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals read from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals with nonmissing phenotypes
Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)
Missing phenotype value is also -9
10 cases, 30 controls and 0 missing
21 males, 19 females, and 0 of unspecified sex
Reading genotype bitfile from [ /opt/galaxy/test-data/tinywga.bed ] 
Detected that binary PED file is v1.00 SNP-major mode
Before frequency and genotyping pruning, there are 25 SNPs
27 founders and 13 non-founders found
Writing allele frequencies (founders-only) to [ tinywga.frq ] 

Analysis finished: Wed May 19 15:15:38 2010


@----------------------------------------------------------@
|        PLINK!       |     v1.06      |   24/Apr/2009     |
|----------------------------------------------------------|
|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |
|----------------------------------------------------------|
|  For documentation, citation & bug-report instructions:  |
|        http://pngu.mgh.harvard.edu/purcell/plink/        |
@----------------------------------------------------------@

Skipping web check... [ --noweb ] 
Writing this text to log file [ tinywga.log ]
Analysis started: Wed May 19 15:15:38 2010

Options in effect:
	--noweb
	--out tinywga
	--bfile /opt/galaxy/test-data/tinywga
	--mind 1.0
	--geno 1.0
	--maf 0.0
	--hwe 0.0
	--missing
	--hardy

Reading map (extended format) from [ /opt/galaxy/test-data/tinywga.bim ] 
25 markers to be included from [ /opt/galaxy/test-data/tinywga.bim ]
Reading pedigree information from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals read from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals with nonmissing phenotypes
Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)
Missing phenotype value is also -9
10 cases, 30 controls and 0 missing
21 males, 19 females, and 0 of unspecified sex
Reading genotype bitfile from [ /opt/galaxy/test-data/tinywga.bed ] 
Detected that binary PED file is v1.00 SNP-major mode
Before frequency and genotyping pruning, there are 25 SNPs
27 founders and 13 non-founders found
Writing Hardy-Weinberg tests (founders-only) to [ tinywga.hwe ] 
0 markers to be excluded based on HWE test ( p <= 0 )
	0 markers failed HWE test in cases
	0 markers failed HWE test in controls
Writing individual missingness information to [ tinywga.imiss ] 
Writing locus missingness information to [ tinywga.lmiss ] 
Total genotyping rate in remaining individuals is 0.995
0 SNPs failed missingness test ( GENO > 1 )
0 SNPs failed frequency test ( MAF < 0 )
After frequency and genotyping pruning, there are 25 SNPs
After filtering, 10 cases, 30 controls and 0 missing
After filtering, 21 males, 19 females, and 0 of unspecified sex

Analysis finished: Wed May 19 15:15:38 2010


@----------------------------------------------------------@
|        PLINK!       |     v1.06      |   24/Apr/2009     |
|----------------------------------------------------------|
|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |
|----------------------------------------------------------|
|  For documentation, citation & bug-report instructions:  |
|        http://pngu.mgh.harvard.edu/purcell/plink/        |
@----------------------------------------------------------@

Skipping web check... [ --noweb ] 
Writing this text to log file [ tinywga.log ]
Analysis started: Wed May 19 15:15:38 2010

Options in effect:
	--noweb
	--out tinywga
	--bfile /opt/galaxy/test-data/tinywga
	--mind 1.0
	--geno 1.0
	--maf 0.0
	--mendel

Reading map (extended format) from [ /opt/galaxy/test-data/tinywga.bim ] 
25 markers to be included from [ /opt/galaxy/test-data/tinywga.bim ]
Reading pedigree information from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals read from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals with nonmissing phenotypes
Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)
Missing phenotype value is also -9
10 cases, 30 controls and 0 missing
21 males, 19 females, and 0 of unspecified sex
Reading genotype bitfile from [ /opt/galaxy/test-data/tinywga.bed ] 
Detected that binary PED file is v1.00 SNP-major mode
Before frequency and genotyping pruning, there are 25 SNPs
27 founders and 13 non-founders found
Total genotyping rate in remaining individuals is 0.995
0 SNPs failed missingness test ( GENO > 1 )
0 SNPs failed frequency test ( MAF < 0 )
After frequency and genotyping pruning, there are 25 SNPs
After filtering, 10 cases, 30 controls and 0 missing
After filtering, 21 males, 19 females, and 0 of unspecified sex
14 nuclear families, 1 founder singletons found
13 non-founders with 2 parents in 13 nuclear families
0 non-founders without 2 parents in 0 nuclear families
10 affected offspring trios
0 phenotypically discordant parent pairs found
Converting data to Individual-major format
Writing all Mendel errors to [ tinywga.mendel ]
Writing per-offspring Mendel summary to [ tinywga.imendel ]
Writing per-family Mendel summary to [ tinywga.fmendel ]
Writing per-locus Mendel summary to [ tinywga.lmendel ]
0 Mendel errors detected in total

Analysis finished: Wed May 19 15:15:38 2010


@----------------------------------------------------------@
|        PLINK!       |     v1.06      |   24/Apr/2009     |
|----------------------------------------------------------|
|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |
|----------------------------------------------------------|
|  For documentation, citation & bug-report instructions:  |
|        http://pngu.mgh.harvard.edu/purcell/plink/        |
@----------------------------------------------------------@

Skipping web check... [ --noweb ] 
Writing this text to log file [ tinywga.log ]
Analysis started: Wed May 19 15:15:38 2010

Options in effect:
	--noweb
	--out tinywga
	--bfile /opt/galaxy/test-data/tinywga
	--mind 1.0
	--geno 1.0
	--maf 0.0
	--check-sex

Reading map (extended format) from [ /opt/galaxy/test-data/tinywga.bim ] 
25 markers to be included from [ /opt/galaxy/test-data/tinywga.bim ]
Reading pedigree information from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals read from [ /opt/galaxy/test-data/tinywga.fam ] 
40 individuals with nonmissing phenotypes
Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)
Missing phenotype value is also -9
10 cases, 30 controls and 0 missing
21 males, 19 females, and 0 of unspecified sex
Reading genotype bitfile from [ /opt/galaxy/test-data/tinywga.bed ] 
Detected that binary PED file is v1.00 SNP-major mode
Before frequency and genotyping pruning, there are 25 SNPs
27 founders and 13 non-founders found
Total genotyping rate in remaining individuals is 0.995
0 SNPs failed missingness test ( GENO > 1 )
0 SNPs failed frequency test ( MAF < 0 )
After frequency and genotyping pruning, there are 25 SNPs
After filtering, 10 cases, 30 controls and 0 missing
After filtering, 21 males, 19 females, and 0 of unspecified sex
Converting data to Individual-major format
Writing X-chromosome sex check results to [ tinywga.sexcheck ] 

Analysis finished: Wed May 19 15:15:38 2010

## Rgenetics: http://rgenetics.org Galaxy Tools rgQC.py Plink pruneLD runner



@----------------------------------------------------------@

|        PLINK!       |     v1.06      |   24/Apr/2009     |

|----------------------------------------------------------|

|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |

|----------------------------------------------------------|

|  For documentation, citation & bug-report instructions:  |

|        http://pngu.mgh.harvard.edu/purcell/plink/        |

@----------------------------------------------------------@



Skipping web check... [ --noweb ] 

Writing this text to log file [ tinywga.log ]

Analysis started: Wed May 19 15:15:38 2010



Options in effect:

	--noweb

	--bfile /opt/galaxy/test-data/tinywga

	--indep-pairwise 40 20 0.5

	--out tinywga



Reading map (extended format) from [ /opt/galaxy/test-data/tinywga.bim ] 

25 markers to be included from [ /opt/galaxy/test-data/tinywga.bim ]

Reading pedigree information from [ /opt/galaxy/test-data/tinywga.fam ] 

40 individuals read from [ /opt/galaxy/test-data/tinywga.fam ] 

40 individuals with nonmissing phenotypes

Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)

Missing phenotype value is also -9

10 cases, 30 controls and 0 missing

21 males, 19 females, and 0 of unspecified sex

Reading genotype bitfile from [ /opt/galaxy/test-data/tinywga.bed ] 

Detected that binary PED file is v1.00 SNP-major mode

Before frequency and genotyping pruning, there are 25 SNPs

27 founders and 13 non-founders found

Total genotyping rate in remaining individuals is 0.995

0 SNPs failed missingness test ( GENO > 1 )

0 SNPs failed frequency test ( MAF < 0 )

After frequency and genotyping pruning, there are 25 SNPs

After filtering, 10 cases, 30 controls and 0 missing

After filtering, 21 males, 19 females, and 0 of unspecified sex

Performing LD-based pruning...

Writing pruned-in SNPs to [ tinywga.prune.in ]

Writing pruned-out SNPs to [ tinywga.prune.out ]

Scanning from chromosome 22 to 22



Scan region on chromosome 22 from [ rs2283802 ] to [ rs4822375 ]

For chromosome 22, 11 SNPs pruned out, 14 remaining



Analysis finished: Wed May 19 15:15:38 2010







@----------------------------------------------------------@

|        PLINK!       |     v1.06      |   24/Apr/2009     |

|----------------------------------------------------------|

|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |

|----------------------------------------------------------|

|  For documentation, citation & bug-report instructions:  |

|        http://pngu.mgh.harvard.edu/purcell/plink/        |

@----------------------------------------------------------@



Skipping web check... [ --noweb ] 

Writing this text to log file [ ldp_tinywga.log ]

Analysis started: Wed May 19 15:15:38 2010



Options in effect:

	--noweb

	--bfile /opt/galaxy/test-data/tinywga

	--extract tinywga.prune.in

	--make-bed

	--out ldp_tinywga



Reading map (extended format) from [ /opt/galaxy/test-data/tinywga.bim ] 

25 markers to be included from [ /opt/galaxy/test-data/tinywga.bim ]

Reading pedigree information from [ /opt/galaxy/test-data/tinywga.fam ] 

40 individuals read from [ /opt/galaxy/test-data/tinywga.fam ] 

40 individuals with nonmissing phenotypes

Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)

Missing phenotype value is also -9

10 cases, 30 controls and 0 missing

21 males, 19 females, and 0 of unspecified sex

Reading genotype bitfile from [ /opt/galaxy/test-data/tinywga.bed ] 

Detected that binary PED file is v1.00 SNP-major mode

Reading list of SNPs to extract [ tinywga.prune.in ] ... 14 read

Before frequency and genotyping pruning, there are 14 SNPs

27 founders and 13 non-founders found

Total genotyping rate in remaining individuals is 0.991071

0 SNPs failed missingness test ( GENO > 1 )

0 SNPs failed frequency test ( MAF < 0 )

After frequency and genotyping pruning, there are 14 SNPs

After filtering, 10 cases, 30 controls and 0 missing

After filtering, 21 males, 19 females, and 0 of unspecified sex

Writing pedigree information to [ ldp_tinywga.fam ] 

Writing map (extended format) information to [ ldp_tinywga.bim ] 

Writing genotype bitfile to [ ldp_tinywga.bed ] 

Using (default) SNP-major mode



Analysis finished: Wed May 19 15:15:38 2010







@----------------------------------------------------------@

|        PLINK!       |     v1.06      |   24/Apr/2009     |

|----------------------------------------------------------|

|  (C) 2009 Shaun Purcell, GNU General Public License, v2  |

|----------------------------------------------------------|

|  For documentation, citation & bug-report instructions:  |

|        http://pngu.mgh.harvard.edu/purcell/plink/        |

@----------------------------------------------------------@



Skipping web check... [ --noweb ] 

Writing this text to log file [ tinywga.log ]

Analysis started: Wed May 19 15:15:38 2010



Options in effect:

	--noweb

	--bfile ldp_tinywga

	--het

	--out tinywga



Reading map (extended format) from [ ldp_tinywga.bim ] 

14 markers to be included from [ ldp_tinywga.bim ]

Reading pedigree information from [ ldp_tinywga.fam ] 

40 individuals read from [ ldp_tinywga.fam ] 

40 individuals with nonmissing phenotypes

Assuming a disease phenotype (1=unaff, 2=aff, 0=miss)

Missing phenotype value is also -9

10 cases, 30 controls and 0 missing

21 males, 19 females, and 0 of unspecified sex

Reading genotype bitfile from [ ldp_tinywga.bed ] 

Detected that binary PED file is v1.00 SNP-major mode

Before frequency and genotyping pruning, there are 14 SNPs

27 founders and 13 non-founders found

Total genotyping rate in remaining individuals is 0.991071

0 SNPs failed missingness test ( GENO > 1 )

0 SNPs failed frequency test ( MAF < 0 )

After frequency and genotyping pruning, there are 14 SNPs

After filtering, 10 cases, 30 controls and 0 missing

After filtering, 21 males, 19 females, and 0 of unspecified sex



 **Warning** this analysis typically requires whole-genome level data

             to give accurate results 



Converting data to Individual-major format

Writing individual heterozygosity information to [ tinywga.het ] 



Analysis finished: Wed May 19 15:15:38 2010





</pre></div></body></html>

<hr>
